### PR TITLE
Polyhedron: add statistics 

### PR DESCRIPTION
**Pull request by maxGimeno**
This PR adds some information to the "Info" dock widget of the polyhedron_items in the Polyhedron_demo:
- presence of self-intersections
- number of degenerated faces (if triangulated)
- number of null length edges
- number of holes
- min/max/mean/mid edge length

All the information is now available for a selection of polyhedron items too.

### CODE REVIEW

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 1]
```diff
@@ -467,12 +480,22 @@ Scene_polyhedron_item::compute_normals_and_vertices(void) const
...
+                vertices[i][1]= p.y();
+                vertices[i][2]= p.z();
+                i = (i+1) %3;
             }
+            if((vertices[0][0] == vertices[1][0] && vertices[0][1] == vertices[1][1] && vertices[0][2] == vertices[1][2])
+            || (vertices[0][0] == vertices[2][0] && vertices[0][1] == vertices[2][1] && vertices[0][2] == vertices[2][2])
+            || (vertices[2][0] == vertices[1][0] && vertices[2][1] == vertices[1][1] && vertices[2][2] == vertices[1][2]))
+                number_of_degenerated_faces++;
+
         }
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: what about using [CGAL::collinear](http://doc.cgal.org/latest/Kernel_23/group__collinear__grp.html) instead to get all degenerated triangles?

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 2]
```diff
@@ -493,6 +516,18 @@ Scene_polyhedron_item::compute_normals_and_vertices(void) const
             positions_lines.push_back(b.z());
             positions_lines.push_back(1.0);
 
+            //statistics
+            edge_lengths.push_back(CGAL::sqrt(
+                                       (a.x()-b.x())*(a.x()-b.x())
+                                      +(a.y()-b.y())*(a.y()-b.y())
+                                      +(a.z()-b.z())*(a.z()-b.z())
+                                       ));
+
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: `CGAL::squared_distance()`

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 3]
```diff
@@ -493,6 +516,18 @@ Scene_polyhedron_item::compute_normals_and_vertices(void) const
...
+            if(a.x() == b.x()
+               && a.y() == b.y()
+               && a.z() == b.z())
+                number_of_null_length_edges++;
+
         }
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: `edge_lengths.back()==0` ?

**Comment by janetournois** [comment]
This PR overlaps with the pull request #405 (already merged)

My strategy was to put the information that is expensive to compute (such as angles) in an external widget and not in "Info" to avoid penalizing the demo runtime when statistics are not needed

**Comment by sloriot** [comment]
I agree with Jane.

File: Three/include/CGAL/Three/Scene_item.h [hunk 4]
```diff
@@ -232,7 +232,19 @@ class SCENE_ITEM_EXPORT Scene_item : public QObject {
...
+  virtual int getnumberOfDegeneratedFaces();
+  //!returns the length of the smallest edge.
+  virtual double getMinEdgesLength();
+  //!returns the length of the biggest edge.
+  virtual double getMaxEdgesLength();
+  //!returns the mean length of the edges.
+  virtual double getMeanEdgesLength();
+  //!returns the median edge.
+  virtual double getMidEdgesLength();*/
 public Q_SLOTS:
```
Comment (lrineau) [inline Three/include/CGAL/Three/Scene_item.h]: I object. Those functions cannot be added to the `Scene_item` class. For several item types, there is not even a notion of "edge". It is completely against the idea of the abstraction `Scene_item`.

File: Polyhedron/demo/Polyhedron/MainWindow.cpp [hunk 5]
```diff
@@ -1241,28 +1241,116 @@ void MainWindow::removeManipulatedFrame(CGAL::Three::Scene_item* item)
...
+      edges += item->polyhedron()->size_of_halfedges()/2;
+      facets += item->polyhedron()->size_of_facets();
+      null_edges += item->getNumberOfNullLengthEdges();
+      null_facet += item->getNumberOfDegeneratedFaces();
+      isolated += item->getNbIsolatedvertices();
+      max_edge = CGAL::max(max_edge, item->getMaxEdgesLength());
+      min_edge = CGAL::min(min_edge, item->getMinEdgesLength());
+      mean += item->getMeanEdgesLength() * item->polyhedron()->size_of_halfedges()/2;
+      total_edges += edges;
+
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/MainWindow.cpp]: All that code should be in the `toolTip()` function of `Scene_polyhedron_item`, and not in the main window.

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp (hunk 3 above)
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: Why ? What does it do ?

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp (hunk 1 above)
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: I'll just use the PMP internal function "is_degenerated" that jane just showed me.

**Comment by maxGimeno** [comment]
Ok, I'll fix what needs to be fixed and move all of it in the "Statistics" widget.

File: Polyhedron/demo/Polyhedron/MainWindow.cpp (hunk 5 above)
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/MainWindow.cpp]: It should be, but I cannot access the info of all selected polyhedrons from here.

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-97

File: Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp [hunk 6]
```diff
@@ -1101,36 +1125,161 @@ void Scene_polyhedron_item::invalidate_aabb_tree()
...
+
+
+  //titles
+  data.titles.append(QString("#Vertices"));
+  data.titles.append(QString("#Facets"));
+  data.titles.append(QString("#Connected Components"));
+  data.titles.append(QString("#Border Edges"));
+  data.titles.append(QString("#Degenerated Faces"));
+  data.titles.append(QString("Connected Components Of The Boundary"));
+  data.titles.append(QString("Area"));
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp]: It should be #Connected Components of the Boundary

File: Three/include/CGAL/Three/Scene_item.h [hunk 7]
```diff
@@ -231,8 +231,21 @@ class SCENE_ITEM_EXPORT Scene_item : public QObject {
...
+   //!Contains the name of the category of statistics and the number of lines it will contain
+   QList<std::pair<QString, int> > categories;
+   //!Contains the name of the lines of each category. Must be sorted from top to bottom.
+   QList<QString> titles;
+  };
+  //!Fills a header_data struct with the header information.
+  virtual void header(header_data&);
+  //!Returns a QString containing the requested value for the the table in the statistics dialog
+  virtual QString compute_stats(int i);
   //!Contrains the number of group and subgroups containing this item.
```
Comment (lrineau) [inline Three/include/CGAL/Three/Scene_item.h]: I do not think that the documentation is sufficient. A reader of that documentation would not know what to write in a new class of scene item.

File: Three/include/CGAL/Three/Scene_item.h [hunk 8]
```diff
@@ -231,8 +231,21 @@ class SCENE_ITEM_EXPORT Scene_item : public QObject {
 
   //!Handles key press events.
   virtual bool keyPressEvent(QKeyEvent*){return false;}
+
+  //!Contains the header for the table in the statistics dialog
+  struct header_data{
+   //!Contains the name of the category of statistics and the number of lines it will contain
```
Comment (lrineau) [inline Three/include/CGAL/Three/Scene_item.h]: Note: The naming convention in CGAL says that class must have the first letter capitalized. In Qt5 the coding 
convention is the the camel-case. That name `header_data` does not fix any convention.

File: Three/include/CGAL/Three/Scene_item.h [hunk 9]
```diff
@@ -231,8 +231,21 @@ class SCENE_ITEM_EXPORT Scene_item : public QObject {
...
+   QList<std::pair<QString, int> > categories;
+   //!Contains the name of the lines of each category. Must be sorted from top to bottom.
+   QList<QString> titles;
+  };
+  //!Fills a header_data struct with the header information.
+  virtual void header(header_data&);
+  //!Returns a QString containing the requested value for the the table in the statistics dialog
```
Comment (lrineau) [inline Three/include/CGAL/Three/Scene_item.h]: The API of that function `header()` is strange. Should not it be the following instead?
```
virtual Header_data header() const;
```

File: Polyhedron/demo/Polyhedron/CMakeLists.txt [hunk 10]
```diff
@@ -186,6 +186,7 @@ if(CGAL_Qt5_FOUND AND Qt5_FOUND AND OPENGL_FOUND AND QGLVIEWER_FOUND)
     Viewer_interface_moc.cpp
     Scene_item_moc.cpp
     Scene_item.cpp #Scene_item.moc
+    Scene_polyhedron_item.cpp
     Scene_group_item.cpp
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/CMakeLists.txt]: That is the error!

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-98

File: Polyhedron/demo/Polyhedron/MainWindow.cpp [hunk 11]
```diff
@@ -1750,3 +1753,97 @@ void MainWindow::recenterSceneView(const QModelIndex &id)
...
+void MainWindow::stat_dlg_update()
+{
+  if(statistics_dlg)
+    statistics_dlg->hide();
+  statistics_on_polyhedron();
+
+}
+void MainWindow::statistics_on_polyhedron()
+{
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/MainWindow.cpp]: Could the name of that function be independent from the item types? The name "polyhedron" should be removed.

File: Polyhedron/demo/Polyhedron/MainWindow.cpp [hunk 12]
```diff
@@ -1750,3 +1753,97 @@ void MainWindow::recenterSceneView(const QModelIndex &id)
...
+  connect(ui.updateButton, SIGNAL(clicked()), this, SLOT(stat_dlg_update()));
+  ui.label_htmltab->setText(get_polyhedron_stats());
+  statistics_dlg->show();
+  statistics_dlg->raise();
+}
+
+/* Creates a string containing an html table. This string is constructed by appending each parts of each row, so that the data can
+  depend on the number of selected items. This String is then returned.*/
+QString MainWindow::get_polyhedron_stats()
+{
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/MainWindow.cpp]: The name "polyhedron" should be removed.

**Comment by sloriot** [comment]
There is one new warning:
`/srv/CGAL/www/Members/testsuite/CGAL-4.8-Ic-98/Polyhedron_Demo/TestReport_afabri_x64_Cygwin-Windows8_MSVC2013-Release-64bits.gz:C:\CGAL\test\CGAL-4.8-Ic-98\cmake\platforms\x64_Cygwin-Windows8_MSVC2013-Release-64bits\test\Polyhedron_Demo\Scene_polyhedron_item.cpp(1190) : warning C4267: 'argument' : conversion from 'size_t' to 'const int', possible loss of data`

**Comment by maxGimeno** [comment]
Fixed

**Pull request merged**